### PR TITLE
Add support to return multiple states

### PR DESCRIPTION
**Pull request by jimthematrix**
This type of return is needed for building a poseidon based cipher, according to this paper https://drive.google.com/file/d/1EVrP3DzoGbmzkRmYnyEDcIQcXVU7GlOd/view.

Equivalent to the latest poseidon functions in circomlibjs: https://github.com/iden3/circomlibjs/blob/main/src/poseidon_wasm.js#L15

### CODE REVIEW

File: poseidon/poseidon.go [hunk 1]
```diff
@@ -63,13 +63,24 @@ func mix(state []*ff.Element, t int, m [][]*ff.Element) []*ff.Element {
...
 // HashWithState computes the Poseidon hash for the given inputs and initState
 func HashWithState(inpBI []*big.Int, initState *big.Int) (*big.Int, error) {
+	res, err := hashWithState(inpBI, initState, 1)
+	if err != nil {
+		return nil, err
+	}
+	return res[0], nil
+}
+
+func hashWithState(inpBI []*big.Int, initState *big.Int, nOuts int) ([]*big.Int, error) {
```
Comment (OBrezhniev) [inline poseidon/poseidon.go:73]: Please rename `hashWithState` to `HashWithStateEx` (and in this way make it public).

File: poseidon/poseidon_test.go [hunk 2]
```diff
@@ -95,6 +95,103 @@ func TestPoseidonHash(t *testing.T) {
...
+	assert.Equal(t,
+		"18586133768512220936620570745912940619677854269274689475585506675881198879027",
+		h[0].String())
+	assert.Equal(t, 1, len(h))
+
+	h, err = HashEx([]*big.Int{b1, b2}, 2)
+	assert.Nil(t, err)
+	assert.Equal(t,
+		"7853200120776062878684798364095072458815029376092732009249414926327459813530",
+		h[0].String())
```
Comment (OBrezhniev) [inline poseidon/poseidon_test.go:128]: Please add checks for actual values for all outputs in these tests. Doesn't have to be so many test cases, just a few to test edge cases and something in the middle. I will add later the same to circomlibjs to compare.

Also please add a few for `HashWithStateEx` if possible.

**Review by OBrezhniev — APPROVED** [review]
